tpm_futurepcr/LogEvent.py

```diff
@@ -185,25 +220,76 @@ def show(self):
-            raise ValueError
+            # Either EFI variables, the ESP, or the .conf, are missing.
+            # It's probably not a systemd-boot environment, so PCR[8] meaning is undefined.
+            logger.verbose("-- not touching non-systemd IPL event --")
+            return self.pcr_extend_values[hash_alg]
+
+        if logger.level <= logging.VERBOSE:
+            old_cmdline = self.data
+            # 2022-03-19 grawity: In the past, we had to strip away the last \0 byte for
+            # some reason (which I don't remember)... but apparently now we don't? Let's
+            # add a warning so that hopefully I remember why it was necessary.
+            if len(old_cmdline) % 2 != 0:
+                logger.warning("Expecting EV_IPL data to contain UTF-16, but length isn't a multiple of 2")
+                old_cmdline += b'\0'
+            old_cmdline = loader_decode_pcr8(old_cmdline)
+            logger.verbose("-- extending with systemd-boot cmdline --")
+            logger.verbose("this cmdline: %s", old_cmdline)
+            logger.verbose("next cmdline: %s", cmdline)
+        cmdline = loader_encode_pcr8(cmdline)
+        return hashlib.new(hash_alg.name.lower(), cmdline).digest()
 
-    def next_extend_value(self, current_extend_value: bytes, hash_alg: str = "sha1") -> bytes:
-        return self._hash_pecoff(hash_alg)
+    def show(self):
+        logger.verbose("\033[1mPCR %d -- Event <%s>\033[m", self.pcr_idx, self.type.name)
+        for i in hexdump(self.data, 64):
+            logger.debug(i)
 
 
-def LogEventFactory(binary_reader, tpm_version, tcg_hdr, substitute_bsa_unix_path, allow_unexpected_bsa) -> BaseEvent | None:
+def LogEventFactory(binary_reader, tpm_version, tcg_hdr, substitute_bsa_unix_path, allow_unexpected_bsa) -> BaseEvent:
     pcr_idx = binary_reader.read_u32()
     type = TpmEventType(binary_reader.read_u32())
 
-    if type == TpmEventType.EFI_BOOT_SERVICES_APPLICATION:
-        return EFIBSAEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type, substitute_bsa_unix_path, allow_unexpected_bsa)
-    else:
-        return GenericEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type)
+    match type:
+        case TpmEventType.EFI_BOOT_SERVICES_APPLICATION:
+            return EFIBSAEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type, substitute_bsa_unix_path, allow_unexpected_bsa)
+        case TpmEventType.EFI_VARIABLE_AUTHORITY | TpmEventType.EFI_VARIABLE_BOOT | TpmEventType.EFI_VARIABLE_DRIVER_CONFIG:
+            return EFIVarEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type)
+        case TpmEventType.IPL:
+            return IPLEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type)
+        case TpmEventType.NO_ACTION:
+            return NoActionEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type)
+        case _:
+            return GenericEvent(binary_reader, tpm_version, tcg_hdr, pcr_idx, type)
```

Comment:
`match` is 3.10, for now I'd like to keep the project compatible with at least python 3.9, since that's what is found in Debian stable and such.

Reply:
ok!